cpp11test/src/test-as.cpp

```diff
@@ -209,6 +239,25 @@ context("as_cpp-C++") {
     UNPROTECT(1);
   }
 
+#if defined(__APPLE__) && defined(R_VERSION) && R_VERSION >= R_Version(3, 5, 0)
+  test_that("as_cpp<std::vector>(ALTREP_INTSXP)") {
+    SEXP r = PROTECT(R_compact_intrange(1, 5));
+
+    expect_true(ALTREP(r));
+    auto x1 = cpp11::as_cpp<std::vector<int>>(r);
+    expect_true(ALTREP(r));
```

Comment:
@jimhester  I really expected *something* here to fail since `as_cpp<IntegerContainer>()` currently relies on `INTEGER()`, how is that working with ALTREP?

Reply:
The vector is still an altrep when you call `INTEGER()` on it, but it is then a _materialized_ altrep vector.

The way to test that you are not materializing an alterep vector is to check that `R_altrep_data2(x) == R_NilValue`

Reply:
e.g. after doing this the second test fails, as you were expecting it to.

```
diff --git a/cpp11test/src/test-as.cpp b/cpp11test/src/test-as.cpp
index 6a3282c..0476ef2 100644
--- a/cpp11test/src/test-as.cpp
+++ b/cpp11test/src/test-as.cpp
@@ -243,9 +243,9 @@ context("as_cpp-C++") {
   test_that("as_cpp<std::vector>(ALTREP_INTSXP)") {
     SEXP r = PROTECT(R_compact_intrange(1, 5));
 
-    expect_true(ALTREP(r));
+    expect_true(R_altrep_data2(r) == R_NilValue);
     auto x1 = cpp11::as_cpp<std::vector<int>>(r);
-    expect_true(ALTREP(r));
+    expect_true(R_altrep_data2(r) == R_NilValue);
 
     expect_true(x1.size() == 5);
     int expected = 1;
```